src/sinks/elasticsearch/common.rs

```diff
@@ -93,6 +93,23 @@ impl ElasticsearchCommon {
 
         let tower_request = config.request.tower.into_settings();
 
+        match (
+            (&config.bulk.version, &config.bulk.version_type),
+            &config.id_key,
+        ) {
+            ((Some(_), VersionType::Internal), _) => {
+                return Err(ParseError::ExternalVersionIgnoredWithInternalVersioning.into());
+            }
+            ((Some(_), VersionType::External), None)
+            | ((Some(_), VersionType::ExternalGte), None) => {
+                return Err(ParseError::ExternalVersioningWithoutDocumentID.into());
+            }
+            ((None, VersionType::External), _) | ((None, VersionType::ExternalGte), _) => {
+                return Err(ParseError::ExternalVersioningWithoutVersion.into());
+            }
+            (_, _) => (),
```

Comment:
How many other cases are there to cover here? By including this total wildcard match arm, we're losing out on any exhaustiveness checking from the compiler that would ensure we're covering all possible cases.

If there are a lot and it would be a pain to include them then maybe that doesn't matter, but the code might be easier to read in that case as three separate `if let` statements to handle the three error cases we're looking for here.

Even better would be to "parse" these three separate fields (which allow representing cases that are not valid, as the potential errors show) into a more specific struct/enum that can only represent valid states (e.g. an `External` variant with an id field, and an `Internal` with no fields), which we could then use without needing any error handling later in the program.

Reply:
Originally I used multiple `if let` conditions but then I wanted to use `match` to catch all possible combinations. But in the end the wildcard makes it useless… I'll switch it back to `if let`. 🙂